error code 401: AUTH_KEY_UNREGISTERED when running updates example

<!-- Please answer these questions before submitting your issue. Thanks! -->

### What version of gotd are you using?

```github.com/gotd/td v0.80.0```

### Can this issue be reproduced with the latest version?
Yes

### What did you do?

Ran the [updates sample](https://github.com/gotd/td/blob/main/examples/updates/main.go)

copy contents to ```updatetest.go```
```go
package main

import (
	"context"
	"os"
	"os/signal"

	"go.uber.org/zap"
	"go.uber.org/zap/zapcore"

	"github.com/gotd/td/telegram"
	"github.com/gotd/td/telegram/updates"
	updhook "github.com/gotd/td/telegram/updates/hook"
	"github.com/gotd/td/tg"
)

func main() {
	ctx, cancel := signal.NotifyContext(context.Background(), os.Interrupt)
	defer cancel()
	if err := run(ctx); err != nil {
		panic(err)
	}
}

func run(ctx context.Context) error {
	log, _ := zap.NewDevelopment(zap.IncreaseLevel(zapcore.InfoLevel), zap.AddStacktrace(zapcore.FatalLevel))
	defer func() { _ = log.Sync() }()

	d := tg.NewUpdateDispatcher()
	gaps := updates.New(updates.Config{
		Handler: d,
		Logger:  log.Named("gaps"),
	})

	os.Setenv("APP_ID", "XXXXX")
	os.Setenv("APP_HASH", "XXXXX")
	os.Setenv("SESSION_FILE", "/path/to/session.json")
	// Initializing client from environment.
	// Available environment variables:
	// 	APP_ID:         app_id of Telegram app.
	// 	APP_HASH:       app_hash of Telegram app.
	// 	SESSION_FILE:   path to session file
	// 	SESSION_DIR:    path to session directory, if SESSION_FILE is not set
	client, err := telegram.ClientFromEnvironment(telegram.Options{
		Logger:        log,
		UpdateHandler: gaps,
		Middlewares: []telegram.Middleware{
			updhook.UpdateHook(gaps.Handle),
		},
	})
	if err != nil {
		return err
	}

	// Setup message update handlers.
	d.OnNewChannelMessage(func(ctx context.Context, e tg.Entities, update *tg.UpdateNewChannelMessage) error {
		log.Info("Channel message", zap.Any("message", update.Message))
		return nil
	})
	d.OnNewMessage(func(ctx context.Context, e tg.Entities, update *tg.UpdateNewMessage) error {
		log.Info("Message", zap.Any("message", update.Message))
		return nil
	})

	return client.Run(ctx, func(ctx context.Context) error {
		// Note: you need to be authenticated here.

		// Fetch user info.
		user, err := client.Self(ctx)
		if err != nil {
			return err
		}
		return gaps.Run(ctx, client.API(), user.ID, updates.AuthOptions{
			OnStart: func(ctx context.Context) {
				log.Info("Gaps started")
			},
		})
	})
}
```

run go file:

```bash
go run updatetest.go
```
Check error:

```bash
2023-04-21T17:19:35.672-0500	INFO	conn.mtproto.rpc	rpc/engine.go:42	Initialized	{"v": "v0.80.0", "conn_id": 0, "dc_id": 2, "retry_interval": "5s", "max_retries": 5}
2023-04-21T17:19:35.672-0500	INFO	telegram/connect.go:114	Starting	{"v": "v0.80.0"}
2023-04-21T17:19:35.673-0500	INFO	telegram/session.go:46	Connection restored from state	{"v": "v0.80.0", "addr": "", "key_id": "XXXXXXX"}
2023-04-21T17:19:35.673-0500	INFO	conn.mtproto.rpc	rpc/engine.go:42	Initialized	{"v": "v0.80.0", "conn_id": 0, "dc_id": 2, "retry_interval": "5s", "max_retries": 5}
2023-04-21T17:19:35.858-0500	INFO	conn.mtproto	mtproto/connect.go:44	Key already exists	{"v": "v0.80.0", "conn_id": 0, "dc_id": 2}
2023-04-21T17:19:36.032-0500	INFO	conn	manager/conn.go:79	SessionInit	{"v": "v0.80.0", "conn_id": 0, "dc_id": 2}
2023-04-21T17:19:36.233-0500	INFO	conn.mtproto.rpc	rpc/engine.go:271	Close called	{"v": "v0.80.0", "conn_id": 0, "dc_id": 2}
2023-04-21T17:19:36.235-0500	INFO	telegram/connect.go:161	Closed	{"v": "v0.80.0"}
panic: callback: rpcDoRequest: rpc error code 401: AUTH_KEY_UNREGISTERED
```


### What did you expect to see?

For the app to print messages

### What did you see instead?

An error

### What Go version and environment are you using?

```
version go1.20.2 darwin/amd64
```

--- a/examples/gif-download/main.go
+++ b/examples/gif-download/main.go
@@ -2,71 +2,29 @@
 package main
 
 import (
-	"bufio"
 	"context"
 	"flag"
 	"fmt"
 	"os"
 	"os/signal"
 	"path/filepath"
-	"strings"
 	"time"
 
 	"github.com/go-faster/errors"
+	"github.com/gotd/contrib/middleware/ratelimit"
 	"go.uber.org/atomic"
 	"go.uber.org/zap"
 	"go.uber.org/zap/zapcore"
-	"golang.org/x/crypto/ssh/terminal"
 	"golang.org/x/sync/errgroup"
 	"golang.org/x/time/rate"
 
-	"github.com/gotd/contrib/middleware/ratelimit"
-
+	"github.com/gotd/td/examples"
 	"github.com/gotd/td/telegram"
 	"github.com/gotd/td/telegram/auth"
 	"github.com/gotd/td/telegram/downloader"
 	"github.com/gotd/td/telegram/query/hasher"
 	"github.com/gotd/td/tg"
 )
-
-// terminalAuth implements auth.UserAuthenticator prompting the terminal for
-// input.
-type terminalAuth struct{}
-
-func (terminalAuth) SignUp(ctx context.Context) (auth.UserInfo, error) {
-	return auth.UserInfo{}, errors.New("not implemented")
-}
-
-func (terminalAuth) AcceptTermsOfService(ctx context.Context, tos tg.HelpTermsOfService) error {
-	return &auth.SignUpRequired{TermsOfService: tos}
-}
-
-func (terminalAuth) Code(ctx context.Context, sentCode *tg.AuthSentCode) (string, error) {
-	fmt.Print("Enter code: ")
-	code, err := bufio.NewReader(os.Stdin).ReadString('\n')
-	if err != nil {
-		return "", err
-	}
-	return strings.TrimSpace(code), nil
-}
-
-func (terminalAuth) Phone(_ context.Context) (string, error) {
-	fmt.Print("Enter phone: ")
-	code, err := bufio.NewReader(os.Stdin).ReadString('\n')
-	if err != nil {
-		return "", err
-	}
-	return strings.TrimSpace(code), nil
-}
-
-func (terminalAuth) Password(_ context.Context) (string, error) {
-	fmt.Print("Enter 2FA password: ")
-	bytePwd, err := terminal.ReadPassword(0)
-	if err != nil {
-		return "", err
-	}
-	return strings.TrimSpace(string(bytePwd)), nil
-}
 
 func run(ctx context.Context) error {
 	var (
@@ -100,7 +58,7 @@ func run(ctx context.Context) error {
 
 	// Setting up authentication flow.
 	// Current flow will read phone, code and 2FA password from terminal.
-	flow := auth.NewFlow(terminalAuth{}, auth.SendCodeOptions{})
+	flow := auth.NewFlow(examples.Terminal{}, auth.SendCodeOptions{})
 
 	// Creating new RPC client.
 	//

--- a/examples/terminal.go
+++ b/examples/terminal.go
@@ -1,0 +1,63 @@
+package examples
+
+import (
+	"bufio"
+	"context"
+	"errors"
+	"fmt"
+	"os"
+	"strings"
+	"syscall"
+
+	"golang.org/x/term"
+
+	"github.com/gotd/td/telegram/auth"
+	"github.com/gotd/td/tg"
+)
+
+// Terminal implements auth.UserAuthenticator prompting the terminal for
+// input.
+//
+// This is only example implementation, you should not use it in your code.
+// Copy it and modify to fit your needs.
+type Terminal struct {
+	PhoneNumber string // optional, will be prompted if empty
+}
+
+func (Terminal) SignUp(ctx context.Context) (auth.UserInfo, error) {
+	return auth.UserInfo{}, errors.New("signing up not implemented in Terminal")
+}
+
+func (Terminal) AcceptTermsOfService(ctx context.Context, tos tg.HelpTermsOfService) error {
+	return &auth.SignUpRequired{TermsOfService: tos}
+}
+
+func (Terminal) Code(ctx context.Context, sentCode *tg.AuthSentCode) (string, error) {
+	fmt.Print("Enter code: ")
+	code, err := bufio.NewReader(os.Stdin).ReadString('\n')
+	if err != nil {
+		return "", err
+	}
+	return strings.TrimSpace(code), nil
+}
+
+func (a Terminal) Phone(_ context.Context) (string, error) {
+	if a.PhoneNumber != "" {
+		return a.PhoneNumber, nil
+	}
+	fmt.Print("Enter phone in international format (e.g. +1234567890): ")
+	phone, err := bufio.NewReader(os.Stdin).ReadString('\n')
+	if err != nil {
+		return "", err
+	}
+	return strings.TrimSpace(phone), nil
+}
+
+func (Terminal) Password(_ context.Context) (string, error) {
+	fmt.Print("Enter 2FA password: ")
+	bytePwd, err := term.ReadPassword(syscall.Stdin)
+	if err != nil {
+		return "", err
+	}
+	return strings.TrimSpace(string(bytePwd)), nil
+}

--- a/examples/updates/main.go
+++ b/examples/updates/main.go
@@ -5,10 +5,13 @@ import (
 	"os"
 	"os/signal"
 
+	"github.com/go-faster/errors"
 	"go.uber.org/zap"
 	"go.uber.org/zap/zapcore"
 
+	"github.com/gotd/td/examples"
 	"github.com/gotd/td/telegram"
+	"github.com/gotd/td/telegram/auth"
 	"github.com/gotd/td/telegram/updates"
 	updhook "github.com/gotd/td/telegram/updates/hook"
 	"github.com/gotd/td/tg"
@@ -31,6 +34,9 @@ func run(ctx context.Context) error {
 		Handler: d,
 		Logger:  log.Named("gaps"),
 	})
+
+	// Authentication flow handles authentication process, like prompting for code and 2FA password.
+	flow := auth.NewFlow(examples.Terminal{}, auth.SendCodeOptions{})
 
 	// Initializing client from environment.
 	// Available environment variables:
@@ -60,13 +66,17 @@ func run(ctx context.Context) error {
 	})
 
 	return client.Run(ctx, func(ctx context.Context) error {
-		// Note: you need to be authenticated here.
+		// Perform auth if no session is available.
+		if err := client.Auth().IfNecessary(ctx, flow); err != nil {
+			return errors.Wrap(err, "auth")
+		}
 
 		// Fetch user info.
 		user, err := client.Self(ctx)
 		if err != nil {
-			return err
+			return errors.Wrap(err, "call self")
 		}
+
 		return gaps.Run(ctx, client.API(), user.ID, updates.AuthOptions{
 			OnStart: func(ctx context.Context) {
 				log.Info("Gaps started")

--- a/examples/userbot/main.go
+++ b/examples/userbot/main.go
@@ -27,6 +27,7 @@ import (
 	"golang.org/x/time/rate"
 	lj "gopkg.in/natefinch/lumberjack.v2"
 
+	"github.com/gotd/td/examples"
 	"github.com/gotd/td/telegram"
 	"github.com/gotd/td/telegram/auth"
 	"github.com/gotd/td/telegram/message/peer"
@@ -211,7 +212,7 @@ func run(ctx context.Context) error {
 	})
 
 	// Authentication flow handles authentication process, like prompting for code and 2FA password.
-	authFlow := auth.NewFlow(terminalAuth{phone: phone}, auth.SendCodeOptions{})
+	flow := auth.NewFlow(examples.Terminal{PhoneNumber: phone}, auth.SendCodeOptions{})
 
 	wg, ctx := errgroup.WithContext(ctx)
 	wg.Go(func() error {
@@ -222,16 +223,9 @@ func run(ctx context.Context) error {
 	wg.Go(func() error {
 		// Spawning main goroutine.
 		if err := client.Run(ctx, func(ctx context.Context) error {
-			if self, err := client.Self(ctx); err != nil || self.Bot {
-				// Starting authentication flow.
-				fmt.Println("Not logged in: starting auth")
-				lg.Info("Starting authentication flow")
-				if err := authFlow.Run(ctx, client.Auth()); err != nil {
-					return errors.Wrap(err, "auth")
-				}
-			} else {
-				fmt.Println("Already logged in")
-				lg.Info("Already authenticated")
+			// Perform auth if no session is available.
+			if err := client.Auth().IfNecessary(ctx, flow); err != nil {
+				return errors.Wrap(err, "auth")
 			}
 
 			// Getting info about current user.

